### ISSUE TITLE
Should have lint to disallow unused type parameters

### ISSUE THREAD

**Issue by lambda** [rust-lang/rust#25871]
The following compiles and runs without warning:

```rust
fn foo<T: Sync> () { }

fn main() {
    foo::<i32>()
}
```

I think that this should at least warn, just like unused imports or unused variables; it would almost always be a result from a mistake in refactoring.

**Comment by oli-obk** [comment]
Note: types already report an error here:

```rust
struct X<T> {
    x: i32,
}

fn main() {}
```

```
<anon>:1:10: 1:11 error: parameter `T` is never used
<anon>:1 struct X<T> {
                  ^
<anon>:1:10: 1:11 help: consider removing `T` or using a marker such as `core::marker::PhantomData`
```

**Comment by shepmaster** [comment]
Dupe of #18549 ?

**Comment by sanxiyn** [comment]
I implemented this in #26684, and there is a question whether the lint should warn-by-default. You may want to comment.

**Comment by steveklabnik** [comment]
Since new lints have a big impact on users of rustc, the policy is that they should go through the RFC process like other user-facing changes. As such, I'm going to give this one a close, but if anyone comes across this ticket and wants this lint, consider adding it to clippy and/or writing up an RFC. Thanks!
